storage/src/test/scala/io/delta/storage/integration/S3LogStoreUtilIntegrationTest.scala

```diff
@@ -0,0 +1,96 @@
+package io.delta.storage.integration
+
+import io.delta.storage.internal.{FileNameUtils, S3LogStoreUtil}
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.fs.s3a.S3AFileSystem
+import org.scalatest.Tag
+import org.scalatest.funsuite.AnyFunSuite
+
+import java.net.URI
+import scala.math.max
+import scala.math.ceil
+import scala.math.round
+
+class S3LogStoreUtilIntegrationTest extends AnyFunSuite {
+  private val runIntegrationTests: Boolean =
+    Option(System.getenv("S3_LOG_STORE_UTIL_TEST_ENABLED")).exists(_.toBoolean)
+  private val bucket = System.getenv("S3_LOG_STORE_UTIL_TEST_BUCKET")
+  private val testRunUID =
+    System.getenv("S3_LOG_STORE_UTIL_TEST_RUN_UID") // Prefix for all S3 keys in the current run
+  private lazy val fs: S3AFileSystem = {
+    val fs = new S3AFileSystem()
+    fs.initialize(new URI(s"s3a://$bucket"), configuration)
+    fs
+  }
+  private val maxKeys = 2
+  private val configuration = new Configuration()
+  configuration.set( // for local testing only
+    "fs.s3a.aws.credentials.provider",
+    "com.amazonaws.auth.profile.ProfileCredentialsProvider",
+  )
+  configuration.set("fs.s3a.paging.maximum", maxKeys.toString)
+
+  private def touch(key: String) {
+    fs.create(new Path(s"s3a://$bucket/$key")).close()
+  }
+
+  private def key(table: String, version: Int): String =
+    s"$testRunUID/$table/_delta_log/%020d.json".format(version)
+
+  private def path(table: String, version: Int): Path =
+    new Path(s"s3a://$bucket/${key(table, version)}")
+
+  private def version(path: Path): Long = FileNameUtils.deltaVersion(path)
+
+  private val integrationTestTag = Tag("IntegrationTest")
+
+  def integrationTest(name: String)(testFun: => Any): Unit =
+    if (runIntegrationTests) test(name, integrationTestTag)(testFun)
+
+  integrationTest("setup delta logs") {
+    touch(s"$testRunUID/empty/some.json")
+    touch(s"$testRunUID/small/_delta_log/%020d.json".format(1))
+    (1 to maxKeys).foreach(v => touch(s"$testRunUID/medium/_delta_log/%020d.json".format(v)))
+    (1 to maxKeys * 10).foreach(v => touch(s"$testRunUID/large/_delta_log/%020d.json".format(v)))
+  }
+
+  integrationTest("empty") {
```

Comment:
empty, small, medium test cases can be refactored to call a helper method which runs the integration test for a given
- path
- list of versions

This would reduce code duplication. Also, is there a reason you only verify the number of list requests for the `large` case?

Reply:
Refactored to share code and to verify the number of list requests for all cases.